rm-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/client/NodeSourceWindow.java

```diff
@@ -0,0 +1,452 @@
+import org.ow2.proactive_grid_cloud_portal.common.client.Images;
+import org.ow2.proactive_grid_cloud_portal.common.client.JSUtil;
+import org.ow2.proactive_grid_cloud_portal.common.client.model.LogModel;
+import org.ow2.proactive_grid_cloud_portal.common.client.model.LoginModel;
+
+import com.google.gwt.core.client.GWT;
+import com.google.gwt.json.client.JSONObject;
+import com.smartgwt.client.types.Alignment;
+import com.smartgwt.client.types.Encoding;
+import com.smartgwt.client.types.FormMethod;
+import com.smartgwt.client.types.Overflow;
+import com.smartgwt.client.widgets.IButton;
+import com.smartgwt.client.widgets.Label;
+import com.smartgwt.client.widgets.Window;
+import com.smartgwt.client.widgets.form.DynamicForm;
+import com.smartgwt.client.widgets.form.fields.CheckboxItem;
+import com.smartgwt.client.widgets.form.fields.FormItem;
+import com.smartgwt.client.widgets.form.fields.HiddenItem;
+import com.smartgwt.client.widgets.form.fields.PasswordItem;
+import com.smartgwt.client.widgets.form.fields.PickerIcon;
+import com.smartgwt.client.widgets.form.fields.SelectItem;
+import com.smartgwt.client.widgets.form.fields.TextItem;
+import com.smartgwt.client.widgets.form.fields.UploadItem;
+import com.smartgwt.client.widgets.layout.HLayout;
+import com.smartgwt.client.widgets.layout.VLayout;
+import com.smartgwt.client.widgets.layout.VStack;
+
+
+public abstract class NodeSourceWindow {
+
+    protected RMController controller;
+
+    protected SelectItem infrastructureSelectItem, policySelectItem;
+
+    protected String previousSelectedInfrastructure = null, previousSelectedPolicy = null;
+
+    protected Window window;
+
+    private String windowTitle;
+
+    private String waitingMessage;
+
+    protected NodeSourceWindow(RMController controller, String windowTitle, String waitingMessage) {
+        this.controller = controller;
+        this.windowTitle = windowTitle;
+        this.waitingMessage = waitingMessage;
+    }
+
+    public void show() {
+        this.window.show();
+    }
+
+    public void destroy() {
+        this.window.destroy();
+    }
+
+    protected abstract boolean isNodeSourceEdited();
+
+    protected abstract void populateFormValues(Label windowLabel, DynamicForm windowForm, TextItem nodeSourceNameItem,
+            CheckboxItem nodesRecoverableItem);
+
+    protected String getPluginShortName(PluginDescriptor plugin) {
+
+        return plugin.getPluginName().substring(plugin.getPluginName().lastIndexOf('.') + 1);
+    }
+
+    protected ArrayList<FormItem> prepareFormItems() {
+
+        infrastructureSelectItem = new SelectItem("infra", "Infrastructure");
+        infrastructureSelectItem.setRequired(true);
+        policySelectItem = new SelectItem("policy", "Policy");
+        policySelectItem.setRequired(true);
+
+        infrastructureSelectItem.setWidth(300);
+        policySelectItem.setWidth(300);
+
+        HiddenItem name = new HiddenItem("nsName");
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Define a constant instead of duplicating this literal "nsName" 3 times. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)